### ISSUE TITLE
Sphinx is unable to find type hints for a `functools.partial`

### ISSUE THREAD

**Issue by elyezer** [sphinx-doc/sphinx#5498]
Subject: Sphinx is unable to find type hints for a `functools.partial`

### Problem
- Running using the nitpick mode makes sphinx build to fail when trying to find type hints for a `functools.partial` object.

#### Procedure to reproduce the problem
```console
$ make html

Running Sphinx v1.8.1
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 67 source files that are out of date
updating environment: 67 added, 0 changed, 0 removed
reading sources... [100%] index                                                                                    

Warning, treated as error:
Invalid type annotation found on functools.partial(<function cli_command at 0x7f280cc96e18>, 'qpc source show'). Ignored: TypeError("functools.partial(<function cli_command at 0x7f280cc96e18>, 'qpc source show') is not a module, class, method, or function.",)
make: *** [Makefile:24: html] Error 2

```

#### Error logs / results
```
Invalid type annotation found on functools.partial(<function cli_command at 0x7f280cc96e18>, 'qpc source show'). Ignored: TypeError("functools.partial(<function cli_command at 0x7f280cc96e18>, 'qpc source show') is not a module, class, method, or function.",)
```

#### Expected results
Sphinx should not raise an warning if it can't find a `functools.partial` or provide an option to ignore such warnings when running on the nitpick mode.

This warning was not raised when running Sphinx 1.7.9

### Reproducible project / your project
- https://github.com/quipucords/camayoc

### Environment info
- OS: Fedora 28
- Python version: 3.6.6
- Sphinx version: 1.8.1

**Comment by tk0miya** [comment]
Note: py3 does not support `typing.get_type_hints(partial_func)`:
```
$ cat test.py
import sys
from functools import partial
from typing import get_type_hints

print('Python version: %s\n' % (sys.version_info,))
def foo(): pass
print(get_type_hints(partial(foo)))
```
```
$ python2.7 test.py
Python version: sys.version_info(major=2, minor=7, micro=15, releaselevel='final', serial=0)

None
```
```
$ python3.5 test.py
Python version: sys.version_info(major=3, minor=5, micro=6, releaselevel='final', serial=0)

Traceback (most recent call last):
  File "test.py", line 7, in <module>
    print(get_type_hints(partial(foo)))
  File "/Users/tkomiya/.pyenv/versions/3.5.6/lib/python3.5/typing.py", line 1544, in get_type_hints
    'or function.'.format(obj))
TypeError: functools.partial(<function foo at 0x10f7278c8>) is not a module, class, method, or function.
```
```
$ python3.6 test.py
Python version: sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)

Traceback (most recent call last):
  File "test.py", line 7, in <module>
    print(get_type_hints(partial(foo)))
  File "/Users/tkomiya/.pyenv/versions/3.6.5/lib/python3.6/typing.py", line 1527, in get_type_hints
    'or function.'.format(obj))
TypeError: functools.partial(<function foo at 0x106136e18>) is not a module, class, method, or function.
```
```
$ python3.7 test.py
Python version: sys.version_info(major=3, minor=7, micro=0, releaselevel='final', serial=0)

Traceback (most recent call last):
  File "test.py", line 7, in <module>
    print(get_type_hints(partial(foo)))
  File "/Users/tkomiya/.pyenv/versions/3.7.0/lib/python3.7/typing.py", line 993, in get_type_hints
    'or function.'.format(obj))
TypeError: functools.partial(<function foo at 0x10a0a71e0>) is not a module, class, method, or function.
```

**Comment by tk0miya** [comment]
Fixed by #5531 .
Thank you for reporting.

**Comment by elyezer** [comment]
Thank you @tk0miya very much for fixing this. I will give it a try later.
